crates/egui/src/widgets/plot/axis.rs

```diff
@@ -0,0 +1,297 @@
+use std::{fmt::Debug, ops::RangeInclusive};
+
+use epaint::{
+    emath::{lerp, remap_clamp, round_to_decimals},
+    Color32, Pos2, Rect, Shape, Stroke, TextShape,
+};
+
+use crate::{Response, Sense, TextStyle, Ui, Widget, WidgetText};
+
+use super::{transform::PlotTransform, GridMark};
+
+pub(super) type AxisFormatterFn = fn(f64, usize, &RangeInclusive<f64>) -> String;
+
+/// Generic constant for x-Axis
+pub(super) const X_AXIS: usize = 0;
+/// Generic constant for y-Axis
+pub(super) const Y_AXIS: usize = 1;
+
+/// Placement of an Axis.
+///
+/// `Default` means bottom for x-axis and left for y-axis.
+/// `Opposite` means top for x-axis and right for y-axis.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Placement {
+    Default,
+    Opposite,
+}
+
+// shorthand types for AxisHints, public API
+/// Configuration for x-axis
+pub type XAxisHints = AxisHints<X_AXIS>;
+/// Configuration for y-axis
+pub type YAxisHints = AxisHints<Y_AXIS>;
```

Comment:
Using const generics like for axis is a bit weird, and will cause code bloat due to monomorpization (basically duplicating all the generated code, increasing e.g. `.wasm` sizes). It would be nice if it could be avoided by passing in the `axis` as an argument/member instead.